EHN Support unit-variance whitening for `FastICA`

jjerphan opened the pull request:
#### Reference Issues/PRs

Supersedes #13096
Fixes #13056

#### What does this implement/fix? Explain your changes.
`FastICA` has an implementation peculiarity: usually, ICA whitening variance is unitary but the current implementation of `FastICA` works with an arbitrary variance for whitening.

This PR:
 - adds the possibility to use unit variance for whitening in `FastICA` using `whiten='unit-variance'`
 - adds the possibility to use arbitrary variance for whitening in `FastICA` using `whiten='arbitrary-variance'`  (the current behaviour)
 - add a deprecation cycle for `whiten`

#### Any other comments?

Currently adapting the PR to address [@jnothman's specs](https://github.com/scikit-learn/scikit-learn/pull/13096#issuecomment-618211854):

---

 - [x] `FastICA()`:
   -   Warn "From version 0.25 whiten='unit-variance' by default" if variance != 1
   -   Compute as in master
 - [x] `FastICA(whiten=True)`:
   -   Warn "From version 0.25 whiten=True should be specified as whiten='arbitrary-variance' (its current behaviour) or whiten='unit-variance'."
   -   compute as in master
 - [x] `FastICA(whiten='unit-variance')`:
   -  no warning
   -  compute with unit variance
 - [x] `FastICA(whiten='arbitrary-variance')`:
   -  no warning
   -  compute as in master

---

fferrin committed:
FIX #13056 - FastICA whitening problem

fferrin committed:
FIX #13056 - FastICA whitening problem: Add suggestions

fferrin committed:
FIX #13056 - FastICA whitening problem: Format

fferrin committed:
FIX #13056 - FastICA whitening problem: Tests

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix warnings

fferrin committed:
FIX #13056 - FastICA whitening problem: Update doc/whats_new

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix conflicts

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix merge conflicts

fferrin committed:
DOC Remove additions to v0.21 change log

fferrin committed:
DOC update v0.22 change log

fferrin committed:
FIX Change log version

fferrin committed:
FIX Remove redundant pytest import

fferrin committed:
Merge branch 'master' into bug-13056

Commit:
Merge branch 'master' into bug-13056

fferrin committed:
Change assert_raises for pytest.raises

fferrin committed:
Merge master and adapt fix to the new release

fferrin committed:
Fix FastICA warns when default value is used

fferrin committed:
Fix test name

fferrin committed:
Fix lint error

fferrin committed:
Delete extra attribute setted in FastICA init

fferrin committed:
Fix variable name

fferrin committed:
Fix mutation of parameter in fit method

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Adapt deprecation cycle tests

jjerphan committed:
Adapt warnings and docstring regarding deprecation

jjerphan committed:
Replace chained np.dot with np.linalg.multi_dot

np.linalg.multi_dot has shown to be faster for the dot
product of 3 matrices.

See:
https://github.com/scikit-learn/scikit-learn/issues/17684#issuecomment-649442725

jjerphan committed:
Reword the whats_new entry

jjerphan committed:
Add an attribute for the effective whitening strategy

jjerphan committed:
Docstring correction

jjerphan committed:
Adapt test for FutureWarning

jjerphan committed:
Adapt the fastica function accordingly

jjerphan committed:
Fix doctest to remove FutureWarning

jjerphan committed:
fixup! Adapt test for FutureWarning

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Simplify returned values

jjerphan committed:
Comply with numpy doc style

jjerphan committed:
fixup! Fix doctest to remove FutureWarning

jjerphan committed:
fixup! Fix doctest to remove FutureWarning

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Ignore UserWarning for test_inverse_transform

jjerphan committed:
Ignore FutureWarning for whiten's new default value in 1.1

jjerphan committed:
Ignore FastICA FutureWarning in test_fit_docstring_attributes

jjerphan committed:
Use valid types for Numpydoc

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Bump versions for deprecation cycles

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan commented:
This is a small gentle up for review, @glemaitre. 🙂

jjerphan commented:
This is a small gentle up for a review, @glemaitre. 🙂

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
GitHub UI is not Julien's friend.

jjerphan committed:
Apply review comments and use 'unit-variance' over 'unitary-variance'

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on sklearn/decomposition/_fastica.py:
```diff
@@ -400,6 +403,9 @@ def my_g(x):
         maximum number of iterations run across all components. Else
         they are just the number of iterations taken to converge.
 
+    _whiten : str or bool
+        The whitening strategy used during fit.
+
```
Now that this is private, we don't need to document it anymore.

```suggestion
```

glemaitre commented on sklearn/decomposition/_fastica.py:
```diff
@@ -585,7 +611,13 @@ def g(x, fun_args):
         self.n_iter_ = n_iter
 
-        if self.whiten:
+        if self._whiten:
+            if self._whiten == "unit-variance":
```
```suggestion
        if self._whiten and self._whiten == "unit-variance":
```

glemaitre commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -289,7 +307,49 @@ def test_fastica_errors():
+    """
+    rng = np.random.RandomState(0)
+    X = rng.random_sample((100, 10))
+    for ica in [FastICA(), FastICA(whiten=True)]:
```
We could parametrize here instead to have an inner loop

glemaitre reviewed (approved):
Otherwise LGTM

jjerphan committed:
Update sklearn/decomposition/_fastica.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Parametrize test_fastica_whiten_default_value_deprecation on ica

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -289,7 +307,49 @@ def test_fastica_errors():
+    with pytest.warns(FutureWarning):
+        Xt = ica.fit_transform(X)
+
+    assert_almost_equal(np.var(Xt), 1.0 / 100)
```
```suggestion
    assert_allclose(np.var(Xt), 1.0 / 100)
```

jjerphan committed:
Use suited interfaces for some assertions

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -289,7 +307,49 @@ def test_fastica_errors():
+    with pytest.warns(FutureWarning):
+        Xt = ica.fit_transform(X)
+
+    assert np.var(Xt) == pytest.approx(1.0 / 100)
```
I think those last two tests could be combined and expanded to check that:

- whiten=True and default value lead to the same result as fitting `whiten='arbitrary-variance'` only using the public API (no need to introspect `ica._whiten` but does not hurt either)
- the first 2 trigger the warning and the latter does not.

jjerphan replied:
I am not sure about what you mean by "the last two tests". Are you referring to this file last test, i.e. `test_fastica_output_shape`? To me, its purposes are somewhat distinct from this test's (i.e `test_fastica_whiten_backwards_compatibility`'s).

d6c4c5e67d2a9fdcf5ec3594a2fda8eb4ec67492 is an attempt to treat your suggestion. Let me know what you think of it.

ogrisel replied:
LGTM, thanks!

ogrisel commented on sklearn/tests/test_docstring_parameters.py:
```diff
@@ -257,6 +257,10 @@ def test_fit_docstring_attributes(name, Estimator):
 
+    # FIXME: TO BE REMOVED for 1.3 (avoid FutureWarning)
+    if Estimator.__name__ == "FastICA":
+        est.set_params(whiten="arbitrary-variance")
```
Better start testing with the future behavior.

```suggestion
        est.set_params(whiten="unit-variance")
```

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -246,18 +262,16 @@ def test_inverse_transform():
+        ("unit-variance", n1): (n_features, n1),
+        ("unit-variance", n2): (n_features, n2),
         (False, n1): (n_features, n2),
         (False, n2): (n_features, n2),
```
This test should also pass with `"arbitrary-variance"` no? (except for the final reversibility check).

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -46,6 +45,10 @@ def test_gs():
+# FIXME remove filter in 1.3
+@pytest.mark.filterwarnings(
+    "ignore:From version 1.3 whiten='unit-variance' will be used by default."
+)
```
Is this really needed? The `whiten` parameter is passed explicitly in this test.

jjerphan replied:
It is needed because "arbitrary-variance" is tested.

ogrisel replied:
I would rather put a local ignore warning context manager where it's actually needed in the test but no big deal if you prefer to keep it this way.

jjerphan replied:
So would I, but there is already a context manager for `ValueErrors` on the call which warns.
Moreover, adding a context on top would just add a fourth indentation level. :/

I think it's fine to keep those 4 lines, which kind of looks like a convention in tests for temporarily filtering warnings.

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -174,6 +179,10 @@ def test_fastica_convergence_fail():
+# FIXME remove filter in 1.3
+@pytest.mark.filterwarnings(
+    "ignore:From version 1.3 whiten='unit-variance' will be used by default."
+)
```
I would rather pass `whiten='unit-variance'` explicitly to the fast-ica call to make sure this test passes with the future behavior and avoid bad surprises later.

ogrisel replied:
If not, we can either adapt the tests or fix `whiten='arbitrary-variance'` right away.

ogrisel commented on sklearn/decomposition/_fastica.py:
```diff
@@ -585,7 +608,13 @@ def g(x, fun_args):
-        if self.whiten:
+        if self._whiten:
+            if self._whiten == "unit-variance":
+                S = np.linalg.multi_dot([W, K, XT]).T
```
If `compute_sources` is True, this is computed twice.

jjerphan replied:
4a4c3c3 horribly solves the problem. Should the logic here be simplified?

ogrisel replied:
I don't have a better suggestion from the top of my head.

ogrisel commented on sklearn/decomposition/_fastica.py:
```diff
@@ -334,9 +329,17 @@ class FastICA(_ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimator)
+    whiten : str or bool, default="warn"
+        Specify the whitening strategy to use.
+        If 'arbitrary-variance' (default), a whitening with variance arbitrary is used.
+        If 'unit-variance', the whitening variance is adjusted to be unitary.
```
Based on my understanding, the following would be both more explicit and correct because I am not sure how "whitening variance" is meant to be defined and I supposed that many readers can be confused as well.

```suggestion
        If 'unit-variance', the whitening matrix is rescaled to ensure that the
        recovered sources each have unit variance.
```

ogrisel reviewed (commented):
Some feedback. Once addressed, LGTM.

jjerphan committed:
Correct deprecation warning and notes

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
DOC Clarify the behavior when `whiten='unit-variance'`

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Use new default for common tests fixture

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Do not recompute S if compute_sources

This adds complexity to the code.
IMO this must be simplified -- I'll come back to it then.

jjerphan committed:
Parametrize test_inverse_transform and test "arbitrary-variance"

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Pass whiten='unit-variance' in test for safety w.r.t future behavior

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Improve backward-compatibility test

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -238,36 +252,39 @@ def test_fit_transform():
 
+    # reversibility test in non-reduction case
+    if n_components == X.shape[1]:
+        assert_array_almost_equal(X, X2)
```
not sure why codecov is complaining here....

jjerphan replied:
Neither do I, though it seems not to complain anymore.

jjerphan replied:
The shape was changed in the previous commit for `12` instead of `10`, making this condition above false.

This is reverted in dc2e42a272

ogrisel commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -238,36 +252,39 @@ def test_fit_transform():
+        ("arbitrary-variance", 10, (12, 10)),
+        ("unit-variance", 5, (12, 5)),
+        ("unit-variance", 10, (12, 10)),
+        (False, 5, (12, 12)),
```
```suggestion
        (False, 5, (12, 5)),
```

to test for reversibility without whitening.

ogrisel commented on sklearn/decomposition/_fastica.py:
```diff
@@ -182,12 +182,18 @@ def fastica(
+        Specify the whitening strategy to use.
+        If 'arbitrary-variance'  (default), a whitening with variance arbitrary is used.
+        If 'unit-variance', the whitening matrix is rescaled to ensure that each
+        recovered sources has unit variance.
```
```suggestion
        recovered source has unit variance.
```

ogrisel reviewed (approved):
I put some comments above, but other than that LGTM!

jjerphan committed:
Fix typos

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Correct test to check for reversibility test

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan commented:
[The sole failing test in the CI](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=34689&view=logs&j=55156e43-4792-506d-3947-cdbc5ace0742&t=6d510dcd-35d3-5892-347a-3b74433a7ac4&l=467) is unrelated to those changes.

jjerphan committed:
Merge branch 'main' into bug-13056

jjerphan committed:
Merge branch 'main' into bug-13056

glemaitre commented on doc/whats_new/v1.1.rst:
```diff
@@ -96,6 +96,13 @@ Changelog
   `Thomas Fan`_.
 
+- |API| :func:`decomposition.FastICA` now supports unit variance for whitening.
+  The default value of its :code:`whiten` argument in will change from
```
If I am not wrong the single backtick should be the same as the `:code:` directive, isn't it?

```suggestion
  The default value of its :code:`whiten` argument will change from
```

jjerphan committed:
DOC Adapt whatsnew entry

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan commented:
I made myself suffer too with this PR. :smile:
